src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,165 @@
+            return (T) dataArray[index];
+        }
         return null;
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (isIndexExist(index)) {
+            dataArray[index] = value;
+        }
     }
 
     @Override
     public T remove(int index) {
+        if (isIndexExist(index)) {
+            return removeElement(index);
+        }
         return null;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (areObjectsEquals(element, dataArray[i])) {
+                return removeElement(i);
+            }
+        }
+        throw new NoSuchElementException("There aren't such elements in the list like " + element);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void growIfFull() {
+        if (dataArray.length == size) {
+            grow(dataArray);
+        }
+    }
+
+    private void grow(Object[] fullArray) {
+        dataArray = new Object[fullArray.length + fullArray.length / 2];
+        System.arraycopy(fullArray, 0, dataArray, 0, fullArray.length);
+    }
+
+    private void growTo(Object[] fullArray, List<T> list) {
+        dataArray = new Object[size + list.size()];
+        System.arraycopy(fullArray, 0, dataArray, 0, fullArray.length);
+    }
+
+    private void addSize() {
+        size++;
+    }
+
+    private void addOnNextFreeSell(T value) {
+        growIfFull();
+        addSize();
+        dataArray[size - 1] = value;
+    }
+
+    private void addByIndex(T value, int index) {
+        growIfFull();
+        System.arraycopy(dataArray, index, dataArray, index + 1, size - index);
+        dataArray[index] = value;
+        addSize();
+    }
+
+    private boolean isIndexExist(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("This index don't exist");
+        } else {
+            return true;
+        }
+    }
+
+    private void indexValidation(int index) {
+        if (index >= 0 && index <= size) {
+            return;
+        }
+        throw new ArrayListIndexOutOfBoundsException("This index don't exist");
+    }
+
+    private boolean areObjectsEquals(T element, Object listElement) {
+        return (element == null && listElement == null)
+                || (element != null && element.equals(listElement));
+    }
+
+    private T removeElement(int index) {
+        indexValidation(index);
+        Object objectToDelete = dataArray[index];
+        if (size - 1 == index) {
+            size--;
+            dataArray[index] = null;
+            return (T) objectToDelete;
+        }
+        size--;
+        System.arraycopy(dataArray, 0, dataArray, 0, index);
+        System.arraycopy(dataArray, index + 1, dataArray, index, size - index);
```

Comment:
I do not think we need 2 `arraycopy` calls here, not sure, but please double-check it

Reply:
In first arraycopy i copy values before index in next i copy values after index and pste them from index position